diff --git a/setupinfo.py b/setupinfo.py
--- a/setupinfo.py
+++ b/setupinfo.py
@@ -114,6 +114,19 @@ def ext_modules(static_include_dirs, static_library_dirs,
     _define_macros = define_macros()
     _libraries = libraries()
 
+    static_libs = []
+    if OPTION_STATIC_DEPS:
+        # Replace -lX flags that bring in shared libraries with libX.a (with path) to link against
+        # their static versions. Note that those .a versions must still be built with -fPIC flag.
+        deps = ['xslt', 'exslt', 'xml2']
+        def find_lib(basename):
+            paths = [os.path.join(d, basename) for d in _library_dirs]
+            return next(p for p in paths if os.path.exists(p))
+        static_libs = [find_lib('lib%s.a' % lib) for lib in deps]
+        _libraries = [lib for lib in _libraries if lib not in deps]
+        # Also specify explicitly additional libraries that lxml's dependencies depend on.
+        _libraries += ['dl', 'pthread', 'rt']
+
     if _library_dirs:
         message = "Building against libxml2/libxslt in "
         if len(_library_dirs) > 1:
@@ -146,7 +159,7 @@ def ext_modules(static_include_dirs, static_library_dirs,
                 sources = [main_module_source],
                 depends = find_dependencies(module),
                 extra_compile_args = _cflags,
-                extra_objects = static_binaries,
+                extra_objects = static_binaries + static_libs,
                 define_macros = _define_macros,
                 include_dirs = _include_dirs,
                 library_dirs = _library_dirs,
@@ -287,13 +300,6 @@ def cflags(static_cflags):
             if not possible_cflag.startswith('-I'):
                 result.append(possible_cflag)
 
-    if sys.platform in ('darwin',):
-        for opt in result:
-            if 'flat_namespace' in opt:
-                break
-        else:
-            result.append('-flat_namespace')
-
     return result
 
 def define_macros():
@@ -446,7 +452,8 @@ OPTION_STATIC = staticbuild or has_option('static')
 OPTION_DEBUG_GCC = has_option('debug-gcc')
 OPTION_SHOW_WARNINGS = has_option('warnings')
 OPTION_AUTO_RPATH = has_option('auto-rpath')
-OPTION_BUILD_LIBXML2XSLT = staticbuild or has_option('static-deps')
+OPTION_STATIC_DEPS = has_option('static-deps')
+OPTION_BUILD_LIBXML2XSLT = staticbuild
 if OPTION_BUILD_LIBXML2XSLT:
     OPTION_STATIC = True
 OPTION_LIBXML2_VERSION = option_value('libxml2-version')
